src/net.md

```diff
@@ -876,6 +877,106 @@ fn run() -> Result<()> {
 # quick_main!(run);
 ```
 
+[ex-check-broken-links]: #ex-check-broken-links
+<a name="ex-check-broken-links"/>
+## Check a webpage for broken links
+
+[![reqwest-badge]][reqwest] [![select-badge]][select] [![url-badge]][url] [![cat-net-badge]][cat-net] |
+
+Use [`Url::parse`] to create a [`Url`] and then create the base URL from that. Use [`reqwest::get`] to perform a HTTP GET request and then use [`Document::from_read`] to parse the response into a HTML document.
+
+We can then check if the document has a "base" tag. If so, we get the "href" [`attr`] from the first occurrence of the "base" tag. This is then used as the base URL.
+
+We can then retrieve all the links from the document by using [`find`] with the criteria of the [`Name`] being "a". 
+This returns a [`Selection`] that we [`filter_map`] on to retrieve the urls from links that have the "href" [`attr`]. We then [`collect`] this into a [`HashSet`] to remove any duplicates.
+
+Now that we have our links, we iterate over each of them and attempt to parse them. If we encounter the [`ParseError::RelativeUrlWithoutBase`] that means the URL is a relative URL, which will need a base, so we add it to the base URL using [`join`]. If any other error is encountered during parsing, we print the error to the console. If the URL is valid, we use [`reqwest::get`] to perform a HTTP GET request for that URL, adding the URL to the "broken_urls" HashSet if response status is 404. If the HashSet is empty, we report that no links are broken, otherwise we iterate over the HashSet and report the broken links.
+
+```rust,no_run
+# #[macro_use]
+# extern crate error_chain;
+extern crate reqwest;
+extern crate select;
+extern crate url;
+
+use url::{Url, ParseError};
+use reqwest::StatusCode;
+use select::document::Document;
+use select::predicate::Name;
+
+use std::collections::HashSet;
+use std::error::Error as StdError;
+#
+# error_chain! {
+#   foreign_links {
+#       ReqError(reqwest::Error);
+#       IoError(std::io::Error);
+#       UrlParseError(url::ParseError);
+#   }
+# }
+
+fn run() -> Result<()> {
+    let mut broken_urls = HashSet::new();
+
+    let url = Url::parse("https://www.rust-lang.org/en-US/")?;
+    let mut base_url = Url::parse(&format!(
+        "{}://{}",
+        url.scheme(),
+        url.host_str().unwrap_or("")
+    ))?;
+
+    let res = reqwest::get(url.as_str())?;
+    let document = Document::from_read(res)?;
+
+    if let Some(base_tag_href) =
+        document
+            .find(Name("base"))
+            .filter_map(|n| n.attr("href"))
+            .nth(0)
+    {
+        base_url = Url::parse(base_tag_href)?;
+    }
+
+    let links: HashSet<&str> = document
+        .find(Name("a"))
+        .filter_map(|n| n.attr("href"))
+        .collect();
+
+    for link in links {
+        let link_url = Url::parse(link).or_else(|e| match e {
```

Comment:
You can use `ParseOptions` here to make the crate do all the work here (if the link is absolut then it will be used otherwise the `base_url` will be used )

```rust
 let base_parser = Url::options().base_url(Some(&base_url));
...
 for link in links { //or `map` here
     let final_url = base_parser.parse(link)?;
```